WordPress/gutenberg
# Block Library: Add Query block.

## Description

This PR introduces a new Query block, which can be described as the block version of the WordPress post loop. Right now, it only supports basic query parameters, but the goal is for it to expose all the relevant parameters of a WP query so that it can be used in any template that needs to render lists of posts.

The implementation has a parent Query block that manages the query and can contain any combination of inner blocks.

Inside the Query block, the Query Loop block reads the query from context and uses it to render the resulting posts using its inner blocks as layout. In the editor, this is achieved by rendering the selected post using `InnerBlocks` and the others using a new live `BlockPreview`. This lets us switch between editing the layout as used by each different post and have it sync across all of the posts.

The Query Pagination block similarly reads the query from context but just uses it to render pagination buttons. In the future, it could be expanded to have its own inner blocks so that you can more freely position the different components of a pagination UI.

All of the new block APIs enabled the implementation to go smoothly. Still, I did have to add a couple of things to the server rendering process: A way to skip inner blocks when rendering a block so that the Query Loop block doesn't recurse infinitely when it has a child Post Content block, and a way to skip dynamic rendering so that the Query Loop block can call its render function without recursing.

This PR is merely the big chunk infrastructure needed for the block, and there is still a lot of UI work to be done in follow-ups. Namely, the design and development of the query toolbar and sidebar controls.

## How to test this?

With the full site editing experiment enabled and some published posts, insert a query block into a new post and play around with the pagination settings while making sure the editor and front end outputs match each other.

Try adding content to the templates and make sure they sync. Also, try moving around the Query Loop and Query Pagination blocks.

## Screenshots

![gif](https://user-images.githubusercontent.com/19157096/81359627-82a79d00-908e-11ea-8a57-035117275f4f.gif)

## Types of Changes

*New Feature:* There is a new set of query (post loop) blocks for full site editing.

## Checklist:
